### PR TITLE
Feature fault tolerance 1.1

### LINKED ISSUE #3774: Move mpConfig-1.3 and mpFaultTolerance-1.1 to July beta
New apis, new OL function for beta.

### PR DESCRIPTION
**Pull request by benjamin-confino**
fixes #3774

### CODE REVIEW

**Comment by benjamin-confino** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.microprofile.config.1.2_fat,com.ibm.ws.microprofile.config.1.2_fat_tck,com.ibm.ws.microprofile.config.1.3_fat,com.ibm.ws.microprofile.config.1.3_fat_tck,com.ibm.ws.microprofile.config_fat,com.ibm.ws.microprofile.config_fat_tck,com.ibm.ws.microprofile.config_git_fat_tck,com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck,com.ibm.ws.microprofile.faulttolerance.cdi_fat,com.ibm.ws.microprofile.faulttolerance.cdi_fat_metrics,com.ibm.ws.microprofile.faulttolerance_fat_tck,com.ibm.ws.microprofile.faulttolerance_git_fat_tck,com.ibm.ws.microprofile.health_fat,com.ibm.ws.microprofile.metrics.1.1_fat_tck,com.ibm.ws.microprofile.metrics_fat,com.ibm.ws.microprofile.metrics_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.microprofile.rest.client_fat,com.ibm.ws.microprofile.rest.client_fat_tck

File: dev/com.ibm.websphere.appserver.features/visibility/public/mpFaultTolerance-1.1/com.ibm.websphere.appserver.mpFaultTolerance-1.1.feature [hunk 1]
```diff
@@ -13,7 +13,7 @@ Subsystem-Name: MicroProfile Fault Tolerance 1.1
  com.ibm.websphere.appserver.mpConfig-1.1; ibm.tolerates:=1.2
 -bundles=com.ibm.ws.require.java8, \
  com.ibm.ws.net.jodah.failsafe.1.0.4; apiJar=false; location:="lib/", \
- com.ibm.ws.microprofile.faulttolerance; apiJar=false; location:="lib/", \
+ com.ibm.ws.microprofile.faulttolerance.1.1; apiJar=false; location:="lib/", \
  com.ibm.ws.microprofile.faulttolerance.spi; apiJar=false; location:="lib/"
 kind=noship
```
Comment (tevans78) [inline dev/com.ibm.websphere.appserver.features/visibility/public/mpFaultTolerance-1.1/com.ibm.websphere.appserver.mpFaultTolerance-1.1.feature]: this should be beta

File: dev/com.ibm.websphere.org.eclipse.microprofile.config.1.3/bnd.bnd [hunk 2]
```diff
@@ -26,7 +26,7 @@ Export-Package: \
   org.eclipse.microprofile.config.spi;version=1.2
```
Comment (tevans78) [inline dev/com.ibm.websphere.org.eclipse.microprofile.config.1.3/bnd.bnd]: should be
   org.eclipse.microprofile.config;version=1.0.1
   org.eclipse.microprofile.config.inject;version=1.0
   org.eclipse.microprofile.config.spi;version=1.3

File: dev/com.ibm.websphere.org.eclipse.microprofile.faulttolerance.1.1/bnd.bnd [hunk 3]
```diff
@@ -24,7 +24,8 @@ Export-Package: \
   org.eclipse.microprofile.faulttolerance;version=1.1, \
   org.eclipse.microprofile.faulttolerance.exceptions;version=1.1
```
Comment (tevans78) [inline dev/com.ibm.websphere.org.eclipse.microprofile.faulttolerance.1.1/bnd.bnd]: exceptions package didn't change, should be 1.0

File: dev/com.ibm.ws.microprofile.faulttolerance.1.1/bnd.bnd [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+# Contributors:
+#     IBM Corporation - initial API and implementation
+#*******************************************************************************
+-include= ~../cnf/resources/bnd/bundle.props
+bVersion=1.0
+
+javac.source: 1.8
+javac.target: 1.8
+
+Bundle-Name: com.ibm.ws.microprofile.faulttolerance.1.1
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.1.1/bnd.bnd]: I don't think this bundle is needed at all, it doesn't add anything, it just copies the contents of the 1.0 version.

File: dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#*******************************************************************************
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/bnd.bnd]: wrong date

File: dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/build.gradle [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/build.gradle]: wrong date

File: dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/fat/src/com/ibm/ws/microprofile/faulttolerance/tck/FaultToleranceTck11Launcher.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/fat/src/com/ibm/ws/microprofile/faulttolerance/tck/FaultToleranceTck11Launcher.java]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/publish/servers/FaultTolerance11TCKServer/bootstrap.properties [hunk 8]
```diff
@@ -0,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/publish/servers/FaultTolerance11TCKServer/bootstrap.properties]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/publish/servers/FaultTolerance11TCKServer/server.xml [hunk 9]
```diff
@@ -0,0 +1,38 @@
+<!--
+    Copyright (c) 2017 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.1.1_fat_tck/publish/servers/FaultTolerance11TCKServer/server.xml]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/AggregatedFTPolicy.java [hunk 10]
```diff
@@ -10,6 +10,8 @@
  *******************************************************************************/
 package com.ibm.ws.microprofile.faulttolerance.cdi;
 
+import java.lang.reflect.Method;
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/AggregatedFTPolicy.java]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceCDIExtension.java [hunk 11]
```diff
@@ -82,7 +82,7 @@ public void beforeBeanDiscovery(@Observes BeforeBeanDiscovery beforeBeanDiscover
         Set<Annotation> annotations = annotatedType.getAnnotations();
         for (Annotation annotation : annotations) {
             //if we find any of the fault tolerance annotations on the class then we will add the intereceptor binding to the class
-            if (FTGlobalConfig.getActiveAnnotations(clazz).contains(annotation.annotationType())) {
+            if (FTGlobalConfig.isAnnotationEnabled(annotation, clazz)) {
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceCDIExtension.java]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceInterceptor.java [hunk 12]
```diff
@@ -111,26 +111,26 @@ private AggregatedFTPolicy processPolicies(InvocationContext context, BeanManage
...
-            } else if (annotation.annotationType().equals(CircuitBreaker.class)) {
-                circuitBreaker = new CircuitBreakerConfig(targetClass, (CircuitBreaker) annotation);
-                circuitBreaker.validate();
-            } else if (annotation.annotationType().equals(Timeout.class)) {
-                timeout = new TimeoutConfig(targetClass, (Timeout) annotation);
-                timeout.validate();
-            } else if (annotation.annotationType().equals(Bulkhead.class)) {
-                bulkhead = new BulkheadConfig(targetClass, (Bulkhead) annotation);
-                bulkhead.validate();
+            //Check that the annotation has not been disabled for this specific class.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceInterceptor.java]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FTGlobalConfig.java [hunk 13]
```diff
@@ -10,10 +10,13 @@
  *******************************************************************************/
 package com.ibm.ws.microprofile.faulttolerance.cdi.config;
 
+import java.lang.annotation.Annotation;
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FTGlobalConfig.java]: date

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FallbackConfig.java [hunk 14]
```diff
@@ -10,6 +10,7 @@
  *******************************************************************************/
 package com.ibm.ws.microprofile.faulttolerance.cdi.config;
 
+import java.lang.reflect.InvocationTargetException;
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FallbackConfig.java]: getting bored of writing "date" ... assume that all the copyright dates need updating

File: dev/com.ibm.ws.microprofile.faulttolerance.metrics/bnd.bnd [hunk 15]
```diff
@@ -37,10 +37,17 @@ src: src,resources
...
         com.ibm.ws.logging.core;version=latest, \
+        com.ibm.ws.kernel.service;version=latest,\
         org.eclipse.osgi;version=latest, \
         com.ibm.websphere.org.osgi.core;version=latest, \
         com.ibm.websphere.org.osgi.service.component;version=latest, \
         com.ibm.wsspi.org.osgi.service.component.annotations;version=latest, \
+        com.ibm.ws.microprofile.faulttolerance;version=latest, \
+        com.ibm.ws.microprofile.faulttolerance.spi;version=latest, \
+        com.ibm.websphere.org.eclipse.microprofile.metrics.1.1.1;version=latest, \
+        com.ibm.websphere.org.eclipse.microprofile.config.1.2.1;version=latest,\
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.faulttolerance.metrics/bnd.bnd]: is 1.2.1 the minimum version of config supported?

Comment (tevans78) [reply dev/com.ibm.ws.microprofile.faulttolerance.metrics/bnd.bnd]: should the package imports specify a range so that config 1.3 and beyond will work?

**Comment by tevans78** [comment]
#build

**Pull request merged**